services/agents/registry.go

```diff
@@ -144,49 +151,43 @@ func (r *Registry) Run(stream agentpb.Agent_ConnectServer) error {
 			err = status.Errorf(codes.Aborted, "Another pmm-agent with ID %q connected to the server.", agent.id)
 			return err
 
-		case msg := <-agent.channel.Requests():
-			if msg == nil {
+		case req := <-agent.channel.Requests():
+			if req == nil {
 				disconnectReason = "done"
 				return agent.channel.Wait()
 			}
 
-			switch req := msg.Payload.(type) {
-			case *agentpb.AgentMessage_Ping:
-				agent.channel.SendResponse(&agentpb.ServerMessage{
-					Id: msg.Id,
-					Payload: &agentpb.ServerMessage_Pong{
-						Pong: &agentpb.Pong{
-							CurrentTime: ptypes.TimestampNow(),
-						},
+			switch p := req.Payload.(type) {
+			case *agentpb.Ping:
+				agent.channel.SendResponse(&channel.Response{
+					ID: req.ID,
+					Payload: &agentpb.Pong{
+						CurrentTime: ptypes.TimestampNow(),
 					},
 				})
 
-			case *agentpb.AgentMessage_StateChanged:
-				if err := r.stateChanged(ctx, req.StateChanged); err != nil {
+			case *agentpb.StateChangedRequest:
+				if err := r.stateChanged(ctx, p); err != nil {
 					l.Errorf("%+v", err)
 				}
 
-				agent.channel.SendResponse(&agentpb.ServerMessage{
-					Id: msg.Id,
-					Payload: &agentpb.ServerMessage_StateChanged{
-						StateChanged: new(agentpb.StateChangedResponse),
-					},
+				agent.channel.SendResponse(&channel.Response{
+					ID:      req.ID,
+					Payload: new(agentpb.StateChangedResponse),
 				})
 
-			case *agentpb.AgentMessage_QanCollect:
-				if err := r.qanClient.Collect(ctx, req.QanCollect.Message); err != nil {
+			case *agentpb.QANCollectRequest:
+				if err := r.qanClient.Collect(ctx, p.Message); err != nil {
 					l.Errorf("%+v", err)
 				}
 
-				agent.channel.SendResponse(&agentpb.ServerMessage{
-					Id: msg.Id,
-					Payload: &agentpb.ServerMessage_QanCollect{
-						QanCollect: new(agentpb.QANCollectResponse),
-					},
+				agent.channel.SendResponse(&channel.Response{
+					ID:      req.ID,
+					Payload: new(agentpb.QANCollectResponse),
 				})
 
-			default:
-				l.Warnf("Unexpected request payload: %s.", msg)
+			case nil:
+				l.Warnf("Unexpected request: %v.", req)
```

Comment:
it makes pmm-managed not backward compatible.

Reply:
You probably mean "forward compatible" – ability to work with newer pmm-agents. pmm-managed is already not forward compatible – if there is an unexpected type, the error would happen earlier, during protobuf decoding. And we don't need to make pmm-managed forward compatible. (We need it in a limited form for pmm-agent.)